### PR TITLE
feat(): auto-reconnect client when closed due to idle connection

### PR DESCRIPTION
**Pull request by crobinson42**
@hyperlink

### CODE REVIEW

**Comment by crobinson42** [comment]
- updated `types/index.d.ts` 
- defaulted `reconnectOnIdle = true` 
- removed package-lock.json, that was sloppy of me to commit, my bad
- changed the test to check that `refreshBroker` called instead of `connect`

Jeeze Travis takes forrrevver to run...

I manually updated the TS defs is that the preferred method for changes or is there some automated script for this? (I don't use TS)

How do you run tests locally?

**Comment by hyperlink** [comment]
> Jeeze Travis takes forrrevver to run...
I noticed this recently. It's never been this slow before...

> How do you run tests locally?

There's a section in the [docs](https://github.com/SOHU-Co/kafka-node#running-tests). It's easy if you're on mac or linux.

> TS defs is that the preferred method for changes or is there some automated script for this? (I don't use TS)

This must be updated manually since we're not using TS but maybe in the future.

Thanks for your contribution!

File: test/test.kafkaClient.js [hunk 1]
```diff
@@ -235,6 +235,19 @@ describe('Kafka Client', function () {
       sandbox.clock.tick();
       sinon.assert.notCalled(client.refreshBrokerMetadata);
     });
+
+    it('should reconnect when broker is closed due to being idle', function () {
+      const client = new Client({ autoConnect: false, reconnectOnIdle: true });
+      const brokerWrapper = client.createBroker('fakehost', 9092, true);
+
+      sandbox.stub(brokerWrapper, 'isIdle').returns(true);
+      sandbox.stub(client, 'reconnectBroker');
```
Comment (hyperlink) [inline test/test.kafkaClient.js]: Should also stub `isValidBroker`

File: test/test.kafkaClient.js [hunk 2]
```diff
@@ -235,6 +235,19 @@ describe('Kafka Client', function () {
...
+      sandbox.useFakeTimers();
+
+      mockSocket.emit('close', false);
+      sandbox.clock.tick(1001);
+      sinon.assert.called(client.reconnectBroker);
```
Comment (hyperlink) [inline test/test.kafkaClient.js]: should also verify called with the `mockSocket`

**Comment by crobinson42** [comment]
@hyperlink tests pass locally with the exception of this "5 pending" line, any thoughts?

Screenshot 2019-03-29 07 39 46

File: test/test.kafkaClient.js (hunk 1 above)
Comment (crobinson42) [reply test/test.kafkaClient.js]: I don't see `isValidBroker` anywhere in the code base, not sure what you're looking for here.

File: test/test.kafkaClient.js (hunk 2 above)
Comment (crobinson42) [reply test/test.kafkaClient.js]: I'm not following you on this comment or what you're looking for here.

File: test/test.kafkaClient.js (hunk 1 above)
Comment (hyperlink) [reply test/test.kafkaClient.js]: https://github.com/SOHU-Co/kafka-node/blob/master/lib/kafkaClient.js#L831

isValidBroker is returning false that's why the travis tests are failing.

File: test/test.kafkaClient.js (hunk 2 above)
Comment (hyperlink) [inline test/test.kafkaClient.js]:
```suggestion
      sinon.assert.calledWithExactly(client.reconnectBroker, mockSocket);
```

File: test/test.kafkaClient.js (hunk 1 above)
Comment (hyperlink) [inline test/test.kafkaClient.js]:
```suggestion
      sandbox.stub(client, 'reconnectBroker');
      sandbox.stub(client, 'isValidBroker').returns(true);
```

Comment (crobinson42) [reply test/test.kafkaClient.js]: Ahh, thx for the line ref link - my forked repo/branch wasn't synced :-( I got it now

File: test/test.kafkaClient.js (hunk 2 above)
Comment (crobinson42) [reply test/test.kafkaClient.js]: `client.reconnectBroker` is being called with `Socket` not `FakeSocket`?

File: test/test.kafkaClient.js [hunk 3]
```diff
@@ -267,6 +267,20 @@ describe('Kafka Client', function () {
       sandbox.clock.tick();
       sinon.assert.notCalled(client.refreshBrokerMetadata);
     });
+
+    it('should reconnect when broker is closed due to being idle', function () {
+      const client = new Client({ autoConnect: false, reconnectOnIdle: true });
+      const brokerWrapper = client.createBroker('fakehost', 9092, true);
+
+      sandbox.stub(brokerWrapper, 'isIdle').returns(true);
+      sandbox.stub(brokerWrapper, 'isValidBroker').returns(true);
```
Comment (hyperlink) [inline test/test.kafkaClient.js]:
```suggestion
      sandbox.stub(client, 'isValidBroker').returns(true);
```

**Comment by hyperlink** [comment]
Thank you for sticking with this!

**Pull request merged**